jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -351,35 +330,67 @@ private ContainerConfigurationTemplate pullContainerConfigJson(
   }
 
   /**
-   * Retrieves the cached base image.
+   * Retrieves the cached base images. If a base image reference is not a manifest list, returns a
+   * single image (if cached). If a manifest list, returns all the images matching the configured
+   * platforms in the manifest list but only when all of the images are cached.
    *
-   * @return the cached image, if found
+   * @return the cached images, if found
    * @throws IOException when an I/O exception occurs
    * @throws CacheCorruptedException if the cache is corrupted
    * @throws LayerPropertyNotFoundException if adding image layers fails
    * @throws BadContainerConfigurationFormatException if the container configuration is in a bad
    *     format
+   * @throws UnlistedPlatformInManifestListException if a cached manifest list has no manifests
+   *     matching the configured platform
    */
-  private Optional<Image> getCachedBaseImage()
+  @VisibleForTesting
+  List<Image> getCachedBaseImages()
       throws IOException, CacheCorruptedException, BadContainerConfigurationFormatException,
-          LayerCountMismatchException {
+          LayerCountMismatchException, UnlistedPlatformInManifestListException {
     ImageReference baseImage = buildContext.getBaseImageConfiguration().getImage();
     Optional<ImageMetadataTemplate> metadata =
         buildContext.getBaseImageLayersCache().retrieveMetadata(baseImage);
     if (!metadata.isPresent()) {
-      return Optional.empty();
+      return Collections.emptyList();
     }
 
+    ManifestTemplate manifestList = metadata.get().getManifestList();
     List<ManifestAndConfigTemplate> manifestsAndConfigs = metadata.get().getManifestsAndConfigs();
-    Verify.verify(manifestsAndConfigs.size() == 1);
-    ManifestTemplate manifest = Verify.verifyNotNull(manifestsAndConfigs.get(0).getManifest());
-    if (manifest instanceof V21ManifestTemplate) {
-      return Optional.of(JsonToImageTranslator.toImage((V21ManifestTemplate) manifest));
+
+    if (manifestList == null) {
+      Verify.verify(manifestsAndConfigs.size() == 1);
+      ManifestTemplate manifest = manifestsAndConfigs.get(0).getManifest();
+      if (manifest instanceof V21ManifestTemplate) {
+        return Collections.singletonList(
+            JsonToImageTranslator.toImage((V21ManifestTemplate) manifest));
+      }
+      return Collections.singletonList(
+          JsonToImageTranslator.toImage(
+              (BuildableManifestTemplate) Verify.verifyNotNull(manifest),
+              Verify.verifyNotNull(manifestsAndConfigs.get(0).getConfig())));
```

Comment:
So this section is for when the base image reference is a single manifest?

Reply:
Exactly. When the image reference is an image manifest and not a manifest list. In the cache metadata, `manifestList` is null in this case.

Reply:
Ah gotcha, is there a particular reason why we chose `null` instead of an empty list for manifestList?

Reply:
Good question. A "manifest list" is not a `List` in Java but a class called `ManifestTemplate`.

Reply:
Right my bad, thanks for clarifying:)